src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,4971 @@
+        {
+            CheckEncodingRules(ruleSet);
+
+            _data = data;
+            _ruleSet = ruleSet;
+        }
+
+        public static bool TryPeekTag(ReadOnlySpan<byte> source, out Asn1Tag tag, out int bytesRead)
+        {
+            return Asn1Tag.TryParse(source, out tag, out bytesRead);
+        }
+
+        public Asn1Tag PeekTag()
+        {
+            if (TryPeekTag(_data.Span, out Asn1Tag tag, out int bytesRead))
+            {
+                return tag;
+            }
+
+            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+        }
+
+        private static bool TryReadLength(
+            ReadOnlySpan<byte> source,
+            AsnEncodingRules ruleSet,
+            out int? length,
+            out int bytesRead)
+        {
+            length = null;
+            bytesRead = 0;
+
+            CheckEncodingRules(ruleSet);
+
+            if (source.IsEmpty)
+            {
+                return false;
+            }
+
+            // T-REC-X.690-201508 sec 8.1.3
+
+            byte lengthOrLengthLength = source[bytesRead];
+            bytesRead++;
+            const byte MultiByteMarker = 0x80;
+
+            // 0x00-0x7F are direct length values.
+            // 0x80 is BER/CER indefinite length.
+            // 0x81-0xFE says that the length takes the next 1-126 bytes.
+            // 0xFF is forbidden.
+            if (lengthOrLengthLength == MultiByteMarker)
+            {
+                // T-REC-X.690-201508 sec 10.1 (DER: Length forms)
+                if (ruleSet == AsnEncodingRules.DER)
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                // Null length == indefinite.
+                return true;
+            }
+
+            if (lengthOrLengthLength < MultiByteMarker)
+            {
+                length = lengthOrLengthLength;
+                return true;
+            }
+
+            if (lengthOrLengthLength == 0xFF)
+            {
+                bytesRead = 0;
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            byte lengthLength = (byte)(lengthOrLengthLength & ~MultiByteMarker);
+
+            // +1 for lengthOrLengthLength
+            if (lengthLength + 1 > source.Length)
+            {
+                bytesRead = 0;
+                return false;
+            }
+
+            // T-REC-X.690-201508 sec 9.1 (CER: Length forms)
+            // T-REC-X.690-201508 sec 10.1 (DER: Length forms)
+            bool minimalRepresentation =
+                ruleSet == AsnEncodingRules.DER || ruleSet == AsnEncodingRules.CER;
+
+            // The ITU-T specifications tecnically allow lengths up to ((2^128) - 1), but
+            // since Span's length is a signed Int32 we're limited to identifying memory
+            // that is within ((2^31) - 1) bytes of the tag start.
```

Comment:
Is that not an issue?

Reply:
In order to work on more than int.MaxValue we'd need to change from Memory/Span to Stream; and then everything would need to allocate.  There's definitely value in the low-allocation version here; if we ever need the >2GB version we'll have to do something else.